libp2p/switch.nim

```diff
@@ -375,17 +375,24 @@ proc stop*(s: Switch) {.async.} =
     warn "error stopping switch", exc = exc.msg
 
 proc subscribeToPeer*(s: Switch, peerInfo: PeerInfo) {.async, gcsafe.} =
+  trace "about to subscribe to pubsub peer", peer = peerInfo.shortLog()
   ## Subscribe to pub sub peer
-  if s.pubSub.isSome and peerInfo.id notin s.dialedPubSubPeers:
+  if s.pubSub.isSome and (peerInfo.id notin s.dialedPubSubPeers):
+    let conn = await s.getMuxedStream(peerInfo)
     try:
-      s.dialedPubSubPeers.incl(peerInfo.id)
-      let conn = await s.dial(peerInfo, s.pubSub.get().codec)
       if isNil(conn):
-        trace "unable to subscribe to peer"
+        trace "unable to subscribe to peer", peer = peerInfo.shortLog
         return
-      await s.pubSub.get().subscribeToPeer(conn)
+
+      s.dialedPubSubPeers.incl(peerInfo.id)
+      if (await s.ms.select(conn, s.pubSub.get().codec)):
+        await s.pubSub.get().subscribeToPeer(conn)
+      else:
+        await conn.close()
```

Comment:
document what raises and what doesn't - every time - else we can read the code

Reply:
for example, looking at this code, I don't know that close doesn't raise - if it does raise, there's a double close here because it catches all exceptions and closes inside the handler - if it doesn't raise, it probably shouldn't be be inside the try block, because it's not one of the statements for which the try blocks applies

Reply:
actually, `close` is guarded and shouldn't raise, this is just precautionary. 

I'm still unsure how to best document throws/no-throws, adding a `# no-throws` comment on every place of invocation? Sounds a bit, erm, dirty? That's my only issue with this annotations/comments, other than that I'm all for it. So if you have some prefered style to annotate this things, I'm all ears...

Reply:
so the thing with precautionary code is that it ends up being confusing, because the reader of the code cannot be sure whether it was precautionary, deliberate or accidental. 

basically, if close doesn't raise, it shouldn't really be in the try block, or at least it should be made explicit that it's there but doesn't raise - with exceptions, one must assume that everything can raise - for example, I think I remember seeing that `close` calls functions which could raise (in mplex for example, it calls `send` which in chronos can raise OSError and whatnot) - I'm not sure how that's changed recently, but it's a possibility that existed before - if the code is annotated with "doesn't raise", the raising is a clear bug that whoever is reading the code can go ahead and fix.

`# no-raises` works for me as annotation - as well as `# raises` on statements that you know can raise - similarly, one can put a comment even on async functions declarations - like `# raises` -  to annotate that they are expected to raise - that'll  give us at least the documentation to start disentangling this jungle.

Reply:
I'm fine with removing the now mostly precautionary `try/except`s for closes. Something to cleanup on the next pass. 

OK, I guess I can always just add a `[# {.raises: Defect.} #]` comment in preparation for a working(TM) `raises`.